release/src/router/perp/perp/man/perpboot.8
.\" perpboot.8
.\" wcm, 2009.11.30 - 2011.02.01
.\" ===
.TH perpboot 8 "January 2013" "perp-2.07" "persistent process supervision"
.SH NAME
perpboot \- startup utility for
.BR perpd (8)
and an associated logger
.SH SYNOPSIS
.B perpboot [\-hV] [\-dx] [
.I basedir
.B ]
.SH DESCRIPTION
.B perpboot
is a utility for starting and monitoring a
.BR perpd (8)
instance on
.I basedir
with an associated logger.
If
.I basedir
is not specified,
.B perpboot
will use the environmental variable PERP_BASE.
If PERP_BASE is not set or empty,
.B perpboot
will use a compiled-in value for
.IR basedir ,
normally
.IR /etc/perp .
.PP
Within
.IR basedir ,
.B perpboot
switches into a subdirectory named
.IR .boot/ .
It then forks two child processes to exec
.I ./rc.log
and
.IR ./rc.perp ,
with a pipe connected between their respective stdin and stdout.
.PP
Normally
.I ./rc.log
will be an executable script that execs into a stdin logger,
such as
.BR tinylog (8)
or
.BR sissylog (8).
Similarly,
.I ./rc.perp
will be an executable script that ultimately execs into
.BR perpd (8).
See the examples section below for sample scripts.
.PP
.B perpboot
sets up its
.I ./rc.log
and
.I ./rc.perp
child processes with an environment that:
.RS
.IP \(bu 4
closes all extraneous file descriptors
.IP \(bu 4
redirects unused stdin, stdout, and stderr to
.I /dev/null
.IP \(bu 4
redirects the stdin of
.I ./rc.log
to the stdout of
.I ./rc.perp
(and vice versa)
.IP \(bu 4
redirects the stderr of
.I ./rc.perp
to stdout (with the effect that both stdout and stderr of
.I ./rc.perp
are captured by the logger)
.IP \(bu 4
sets the environmental variable PERP_BASE to the value used for
.I basedir
.RE
.PP
By default,
.B perpboot
itself acts as lightweight supervisor for the
.I ./rc.log
and
.I ./rc.perp
child processes.
If
.B perpboot
notices that either of these processes has died,
it will try to restart them.
.SH OPTIONS
.TP
.B \-d
Detach.
Normally
.B perpboot
itself runs in the foreground.
The
.B \-d
option causes
.B perpboot
to detach from the controlling terminal and run as a background process.
This option is useful when starting
.B perpboot
from within a BSD-type boot script
such as
.IR rc.local .
.TP
.B \-h
Help.
Print a brief usage message to stderr and exit.
.TP
.B \-V
Version.
Print the version number to stderr and exit.
.TP
.B \-x
Exit.
Normally
.B perpboot
stays resident as a system process that monitors its
.I ./rc.log
and
.I ./rc.perp
child processes.
The
.B \-x
option causes
.B perpboot
to start
.I ./rc.log
and then replace itself with the
.I ./rc.perp
process.
This option is used to reduce process overhead
and/or when you don't feel the need to retain
.B perpboot
as a supervisor.
As one example,
this option might be used within an
.BR inittab (5)
specification configured with the
.I action
specified as ``respawn''.
.SH EXAMPLES
An example
.I ./rc.log
script may look like this:
.PP
.RS
.nf
#!/bin/sh
LOGDIR=/var/log/perpd
exec tinylog -k 8 -s 100000 -t ${LOGDIR}
.fi
.RE
.PP
This script execs into an instance of
.BR tinylog (8)
maintaining a rotated set of log files in
.I /var/log/perpd
with a maximum of 8 keep files,
a maximum log size of 100000 bytes,
and with timestamp prepended entries.
.PP
An example
.I ./rc.perp
script:
.PP
.RS
.nf
#!/bin/sh
exec perpd -a 6 $PERP_BASE
.fi
.RE
.PP
Here the script execs into
.BR perpd (8)
with autoscanning set to 6 seconds.
Note also that the environmental variable PERP_BASE is defined and available for the
.I ./rc.perp
and
.I ./rc.log
scripts.
.PP
These simple scripts could be customized and embellished considerably.
For example, the
.I ./rc.log
script could:
.RS
.IP \(bu 4
check/create existence/permissions of log directory
.IP \(bu 4
drop privilege before running
.BR tinylog (8)
\-\- see
.BR runuid (8)
.RE
.PP
Similarly, the
.I ./rc.perp
script could perform other initializations before the
.BR perpd (8)
exec:
.RS
.IP \(bu 4
check/create existence and symlink for
.I /etc/perp/.control
.IP \(bu 4
clean out stale control directories/files in
.I /etc/perp/.control
.RE
.PP
.B perpd
itself may be invoked in the way that best suits the
.BR init (8)
system and boot scripts of the host environment.
A sysv-compatible system may use an
.BR inittab (5)
file configured with this entry:
.PP
.RS
.nf
PB:23456:respawn:/usr/sbin/perpboot -x /etc/perp
.fi
.RE
.PP
This example shows the
.B \-x
option,
with the effect that
.BR init (8)
itself will monitor the
.BR perpd (8)
process and respawn it if it dies.
.PP
A BSD-type system may use an
.I rc.local
file with this entry:
.PP
.RS
.nf
if [ -x /usr/sbin/perpboot ]; then
    echo -n ' perpd'
      /usr/sbin/runenv -i /etc/perp/.boot/perp.env \\\^ 
          /usr/sbin/perpboot -d /etc/perp
fi
.fi
.RE
.PP
In this example, the
.B \-d
option is used to run
.B perpboot
as a daemon process.
.PP
This example is further embellished to show the use of the
.BR runenv (8)
runtool to setup a clean environment for the
.BR perpd (8)
process and its children.
.SH FILES
.I /etc/perp/
.RS
The default base operating directory for
.BR perpd (8).
.RE
.PP
.I /etc/perp/.boot/
.RS
Directory containing the startup scripts used by
.BR perpboot .
.RE
.PP
.I /etc/perp/.boot/rc.log
.RS
Control script used by
.B perpboot
to start a logger for
.BR perpd (8).
.RE
.PP
.I /etc/perp/.boot/rc.perp
.RS
Control script used by
.B perpboot
to start up
.BR perpd (8).
.RE
.SH ENVIRONMENT
PERP_BASE
.RS
If no
.I basedir
argument is given at the command-line on startup,
.B perpboot
checks for a value defined by PERP_BOOT.
If this is not defined or empty,
.B perpboot
uses a compiled-in default, normally
.IR /etc/perp .
.SH SIGNALS
If
.B perpboot
gets the SIGTERM signal,
it performs a shutdown sequence on its child processes:
.RS
.IP \(bu 4
sends SIGTERM and SIGCONT to the
.I ./rc.perp
process and waits for it to terminate
.IP \(bu 4
closes its own copy of the input pipe to the
.I ./rc.log
process and waits for the logger to terminate
.IP \(bu 4
exits 0
.RE
.PP
Otherwise, when
.B perpboot
runs as a lightweight supervisor
(without the
.B \-x
option),
it traps all the other signals it can and relays them directly to the
.I ./rc.perp
child process with
.BR kill (2).
.SH NOTES
The
.B perpboot
utility is a purpose-specific modification of
.BR rundeux (8).
.SH AUTHOR
Wayne Marshall, http://b0llix.net/perp/
.SH SEE ALSO
See the
.BR perp-setup (8)
utility for an ``automagic'' configuration of a
.B perpboot
installation.
.PP
.nh
.BR perp_intro (8),
.BR perp-setup (8),
.BR perpctl (8),
.BR perpd (8),
.BR perpetrate (5),
.BR perphup (8),
.BR perpls (8),
.BR perpok (8),
.BR perpstat (8),
.BR sissylog (8),
.BR tinylog (8),
.BR rundeux (8)
.\" EOF perpboot.8


release/src/router/softcenter_ext/perp/perp/perpboot.c
/* perpboot.c
** perp: persistent process supervision
** perp 2.0: single process scanner/supervisor/controller
** perpboot: perpd startup utility
**   - exec and supervise perpd + logger
**   - a lightweight process supervisor for perpd itself
** (see also rundeux(8) in runtools)
** wcm, 2009.10.04 - 2011.03.22
** ===
*/

/* TODO
**   * built-in envfile processing
*/

#include <stdlib.h>
/* unix: */
#include <unistd.h>
#include <errno.h>
#include <fcntl.h>
#include <poll.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/resource.h>
#include <sys/wait.h>

/* lasagna: */
#include "cstr.h"
#include "fd.h"
#include "newenv.h"
#include "nextopt.h"
#include "sig.h"
#include "sigset.h"
#include "tain.h"

#include "perp_common.h"
#include "perp_stderr.h"

/* other macros: */
#define warn(...) \
    eputs(progname, ": warning: ", __VA_ARGS__)

#define do_kill(p,k) \
    if((p) > 0) kill((p),(k))

/* environ: */
extern char  **environ;

static const char *progname = NULL;
static const char prog_usage[] = "[-hV] [-d] [-x] [dir]";

struct child {
  char     *argv[2];
  pid_t     pid;
  tain_t    when;
};

/* variables in scope: */
/* deux[0] = logger, deux[1] = program: */
static struct child  deux[2];
static int           my_sigpipe[2];
static int           my_logpipe[2];
static int           flag_term = 0;
static sigset_t      my_sigset;
static char         *perp_base = NULL;
/* running as supervisor (by default): */
static int           opt_super = 1;

static void sigtrap(int sig);
static void sigpipe_ping(void);
static void setup(void);
static void deux_init(void);
static void child_exec(int which);
static void child_wait(void);
static void main_loop(void);


static
void
sigtrap(int sig)
{
  int  want_cont = 0;

  switch(sig){
  case SIGCHLD:
      break;
  case SIGINT:
      /* want clean term with SIGINT! */
      /* fall through: */
  case SIGTERM:
      flag_term = 1;
      want_cont = 1;
      /* fallthrough: */
  default:
      /* by default, pass signals to perpd: */
      do_kill(deux[1].pid, sig);
      if(want_cont){
          do_kill(deux[1].pid, SIGCONT);
      }
      break;
  }

  sigpipe_ping();
  return;
}


static
void
sigpipe_ping(void)
{
  int  terrno = errno;
  int  w;

  do{
      w = write(my_sigpipe[1], "!", 1);
  }while((w == -1) && (errno == EINTR));

  errno = terrno;
  return;
}


static
void
setup(void)
{
  int  i = 0;

  if(pipe(my_sigpipe) == -1){
      fatal_syserr("failure setting up selfpipe");
  }

  for(i = 0; i < 2; ++i){
      fd_cloexec(my_sigpipe[i]);
      fd_nonblock(my_sigpipe[i]);
  }

  if(pipe(my_logpipe) == -1){
      fatal_syserr("failure creating logpipe");
  }

  sigset_fill(&my_sigset);
  sigset_block(&my_sigset);

  sig_catch(SIGTERM, &sigtrap);
  sig_catch(SIGINT, &sigtrap);
  sig_catch(SIGCHLD, &sigtrap);

  /* catching these signals to pass to perpd: */
  sig_catch(SIGALRM, &sigtrap);
  sig_catch(SIGCONT, &sigtrap);
  sig_catch(SIGHUP, &sigtrap);
  sig_catch(SIGQUIT, &sigtrap);
  sig_catch(SIGTSTP, &sigtrap);
  sig_catch(SIGUSR1, &sigtrap);
  sig_catch(SIGUSR2, &sigtrap);

  sig_ignore(SIGPIPE);
  return;
}    


static
void
deux_init(void)
{
  /* deux[0]: logger: */
  deux[0].pid = 0;
  tain_load(&deux[0].when, 0, 0);
  deux[0].argv[0] = "./rc.log";
  deux[0].argv[1] = NULL;

  /* deux[1]: program: */
  deux[1].pid = 0;
  tain_load(&deux[1].when, 0, 0);
  deux[1].argv[0] = "./rc.perp";
  deux[1].argv[1] = NULL;

  return;
}


static
void
child_exec(int which)
{
  pid_t      pid = 0;
  tain_t     now;
  tain_t     when_ok;
  char     **argv = deux[which].argv;

  tain_now(&now);
  tain_load(&when_ok, 1, 0);
  tain_plus(&when_ok, &deux[which].when, &when_ok);
  if(tain_less(&now, &when_ok)){
      warn("pausing for restart of ", argv[0], " ...");
      sleep(1);
  }

  /*  if option -x:
  **    not supervising, don't fork here, just exec into deux[1]
  */
  if(opt_super ? 1 : (which == 0)){
      while((pid = fork()) == -1){
          warn("failure on fork() while starting ", argv[0]);
          sleep(2);
      }
  }
      
  /* child (or, if not opt_super, execing into deux[1]): */
  if(pid == 0){
      int            fd, fd_max;
      struct rlimit  rlim;
      int            i;

      /* setup PERP_BASE in environment: */
      if(newenv_set("PERP_BASE", perp_base) == -1){
          fatal_syserr("failure setting environment in child for ", argv[0]);
      }
      /* prepare for closing unused file descriptors: */
      if(getrlimit(RLIMIT_NOFILE, &rlim) == -1){
          fatal_syserr("failure getting file rlimit in child for ", argv[0]);
      }
      fd_max = (rlim.rlim_max == RLIM_INFINITY) ? 1024 : rlim.rlim_max;
      /* start fd 0,1,2 on /dev/null: */
      if((fd = open("/dev/null", O_RDWR)) == -1){
          fatal_syserr("failure opening /dev/null in child for ", argv[0]);
      }
      fd_dupe(0, fd); fd_dupe(1, fd); fd_dupe(2, fd);
      close(fd);
      /* setup logpipe: */
      fd_dupe(which, my_logpipe[which]);
      if(which == 1){
          /* perpd gets stderr redirected to stdout for logger: */
          fd_dupe(2, 1);
      }
      /* close all other descriptors: */
      for(i = 3; i < fd_max; ++i) close(i);
      /* set default umask: */
      umask(0);
      /* reset default signal handlers, unblock: */
      sig_default(SIGTERM);
      sig_default(SIGCHLD);
      sig_default(SIGALRM);
      sig_default(SIGCONT);
      sig_default(SIGHUP);
      sig_default(SIGINT);
      sig_default(SIGQUIT);
      sig_default(SIGTSTP);
      sig_default(SIGUSR1);
      sig_default(SIGUSR2);
      sig_default(SIGPIPE);
      sigset_unblock(&my_sigset);
      /* do it: */
      newenv_run(argv, environ);
      /* uh oh: */
      fatal_syserr("failure on exec of ", argv[0]);
  }

  /* parent: */
  deux[which].pid = pid;
  tain_now(&deux[which].when);

  return;
}


static
void
child_wait(void)
{
  pid_t  pid;

  while((pid = waitpid(-1, NULL, WNOHANG)) > 0){
      if(pid == deux[0].pid){
          deux[0].pid = 0;
      }else if(pid == deux[1].pid){
          deux[1].pid = 0;
      }
  }

  return;
}


static
void
main_loop(void)
{
  struct pollfd  pollv[1];
  int            i;
  int            e = 0;
  char          *c;

  pollv[0].fd = my_sigpipe[0];
  pollv[0].events = POLLIN;

  for(;;){
      if(flag_term){
          break;
      }

      /* start/restart children: */
      /* deux[0] is logger, check it first: */
      for(i = 0; i < 2; ++i){
          if(deux[i].pid == 0){
              child_exec(i);
          }
      }

      /* poll on sigpipe: */
      sigset_unblock(&my_sigset);
      do{
          e = poll(pollv, 1, -1);
      }while((e == -1) && (errno == EINTR));
      sigset_block(&my_sigset);

      /* consume sigpipe: */
      while(read(my_sigpipe[0], &c, 1) == 1){/*empty*/ ;}

      /* consume dead children: */
      child_wait();
  }

  /* here on SIGTERM: */
  if(deux[1].pid){
      /* harvest deux[1]: */
      waitpid(deux[1].pid, NULL, 0);
  } 
  if(deux[0].pid){
      /* logger should exit on eof: */
      close(my_logpipe[1]);
      do_kill(deux[0].pid, SIGCONT);
      waitpid(deux[0].pid, NULL, 0);
  }

  return;
}


int
main(int argc, char *argv[])
{
  nextopt_t  nopt =  nextopt_INIT(argc, argv, ":hVdx");
  char       opt;
  int        opt_detach = 0;

  progname = nextopt_progname(&nopt);
  while((opt = nextopt(&nopt))){
      char optc[2] = {nopt.opt_got, '\0'};
      switch(opt){
      case 'V': version(); die(0); break;
      case 'h': usage(); die(0); break;
      case 'd': opt_detach = 1; break;
      case 'x': opt_super = 0; break;
      case ':':
          fatal_usage("missing argument for option -", optc);
          break;
      case '?':
          if(nopt.opt_got != '?'){
              fatal_usage("invalid option: -", optc);
          }
          /* else fallthrough: */
      default :
          die_usage(); break; 
      }
  }

  argc -= nopt.arg_ndx;
  argv += nopt.arg_ndx;

  perp_base = argv[0];

  /* perp_base: default is "/etc/perp": */
  if((perp_base == NULL) || (perp_base[0] == '\0')){
      perp_base = getenv("PERP_BASE");
  }
  if((perp_base == NULL) || (perp_base[0] == '\0')){
      perp_base = PERP_BASE_DEFAULT;
  }

  /* initialize sigpipe[], etc: */
  setup();

  if(opt_detach){
      pid_t  pid = fork();
      switch(pid){
      case -1: fatal_syserr("failure fork(): unable to detach"); break;
      case 0:
          /* child continues daemonized in new session: */
          setsid(); break;
      default:
          /* parent exits: */
          exit(0);
      }
  }

  if(chdir(perp_base) == -1){
      fatal_syserr("failure chdir() to ", perp_base);
  }
  if(chdir(PERP_BOOT) == -1){
      fatal_syserr("failure chdir() to ", perp_base, "/", PERP_BOOT);
  }

  deux_init();
  main_loop();

  return 0;
}


/* eof: perpboot.c */
